### ISSUE TITLE
instance Enum IoSubSystem is missing in GHC 9.12-alpha

### ISSUE THREAD

**Issue by tomjaguarpaw** [haskell/core-libraries-committee#311]
Adithya Kumar [wrote to the `ghc-devs` list to report](https://mail.haskell.org/pipermail/ghc-devs/2024-December/021867.html) that `instance Enum IoSubSystem` is missing in GHC 9.12-alpha. I believe this change was not approved by the CLC. It looks like that change snuck in despite the regression tests in place to catch changes to `base`'s API. See [my reply](https://mail.haskell.org/pipermail/ghc-devs/2024-December/021873.html) for more details on the matter.

How should the CLC proceed, if indeed the change was not approved?

**Comment by hasufell** [comment]
> How should the CLC proceed, if indeed the change was not approved?

It must be reverted.

**Comment by tomjaguarpaw** [comment]
What's the process to achieve that? Who do we notify? Who prepares the patch?

**Comment by tomjaguarpaw** [comment]
OK, [I've notified the `ghc-devs` list](https://mail.haskell.org/pipermail/ghc-devs/2024-December/021874.html) that the opinion of the CLC will likely be that the change should be reverted before release.

@Bodigrim: how do we officially decide on this matter and ensure the correct process is followed in consequence?

**Comment by TeofilC** [comment]
By revert do y'all mean the entire patch or the removal of the instances?

**Comment by tomjaguarpaw** [comment]
The removal of the instance. It's not CLC's place to determine *how* the API is preserved, just that it _is_ preserved.

**Comment by bgamari** [comment]
I did some digging to sort out what happened here. Unfortunately it appears that the situation is a bit worse than just an `Enum` instance. However, I think it can be remediated, as I will describe below.

# Background

The cited commit makes the following publicly visible changes in `GHC.RTS.Flags`:

 1. the `getIoManagerFlag :: IO IoSubSystem` function was removed
 2. the `IoSubSystem` type is no longer exported from `GHC.RTS.Flags`
 3. the `IoManagerFlag` type is now exported
 4. the type of `MISC_FLAGS`'s `ioManager` field has changed from `IoSubSystem` to `IoManagerFlag`.

These changes happened as a side-effect of a rather intricate set of refactorings carried out in !9676, where the RTS's logic for the selection of IO manager backends was rewritten. Change (1) was the removal of a function introduced for internal use which likely should never have been public to begin with (today it would have been available only via `ghc-internal`). Changes (2), (3), and (4) update the Haskell interface to reflect the current RTS behavior.

Unfortunately, the `base-exports` test changes were accepted during a post-review rebase (as the test did not exist at the time that this MR was posted) and consequently slipped under the radar.

# Proposed course of action

As !9676 has been in the works for multiple years and merged over 8 months ago I think we should be quite reluctant to revert it wholesale. Reverting after 8 months of development has occurred on top of the new status quo would likely be a delicate operation in a rather subtle area of the compiler.

However, my sense is that we might be able to revert the user-visible changes in `GHC.RTS.Flags` by approximating the old Haskell interfaces on top of the new behavior. Specifically,

To mitigate change (1) I would recommend that we reintroduce `getIoManagerFlag` as

```haskell
getIoManagerFlag :: IO IoSubSystem
getIoManagerFlag = return GHC.Internal.IO.SubSystem.ioSubSystem
```

Similarly, we migitate (4) by reverting the change in type of the `ioManager` field and modifying `getMiscFlags` to fill it with `GHC.Internal.IO.SubSystem.ioSubSystem`.

The export removed in change (2) would be reinstated (its definition having been moved to `GHC.Internal.IO.SubSystem`.

The export introduced in (3) would be removed.

This would restore things to the state they had in 9.10. In 9.12 we will likely propose the removal of all affected interfaces as `GHC.RTS.Flags` will be frozen under core-libraries-committee#289.

**Comment by Bodigrim** [comment]
> In 9.12 we will likely propose the removal of all affected interfaces as `GHC.RTS.Flags` will be frozen under core-libraries-committee#289.

9.12 or 9.14?

**Comment by bgamari** [comment]
Sigh, yes, indeed I intended 9.14. I have amended the comment.

**Comment by Bodigrim** [comment]
> @Bodigrim: how do we officially decide on this matter and ensure the correct process is followed in consequence?

If changes have not been released (as in this case), there is nothing particular for us to decide. Anyone (including CLC members of course) can raise the violation to the attention of GHC team and ask to revert user-visible changes.

When I spot it I usually leave a message in the MR which introduced accidental unapproved breakage. An author is likely interested to avoid complete revert and also is in the best position to carefully craft a partial revert.

If a GHC release is close (as in this case) I'd also CC a release manager, because it's a blocker and needs to be rectified and backported quickly. 

CC @wz1000 @bgamari @dcoutts 
Guys, given that it was a big and complex MR, I don't think it's a good idea for outsiders to attempt a partial revert, especially at this timescale. Could you possibly rectify the matters by restoring API as it was in GHC 9.10?

**Comment by bgamari** [comment]
@Bodigrim I have described one mitigation in my comment [above](https://github.com/haskell/core-libraries-committee/issues/311#issuecomment-2518418231). I have opened [!13683](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/13683) to validate that this approach is practical.

**Comment by AndreasPK** [comment]
It seems this approved CLC proposal flew under the rader in this discussion: https://github.com/haskell/core-libraries-committee/issues/263

**Comment by bgamari** [comment]
> It seems this approved CLC proposal flew under the rader in this discussion: #263

I'm not sure how I overlooked this; the changes are even noted in the changelog.

In this case my [comment](https://github.com/haskell/core-libraries-committee/issues/311#issuecomment-2518418231) can be ignored. We can merely reintroduce the missing instances (`Show IoSubSystem` is also missing).

**Comment by dcoutts** [comment]
@tomjaguarpaw says ([on the mailing list](https://mail.haskell.org/pipermail/ghc-devs/2024-December/021873.html))

> There is test in place to confirm that the API of base does not change accidentally.  The test correctly caught the removal of this [instance](https://gitlab.haskell.org/ghc/ghc/-/commit/1adc6fa43e0c2f577c15f560d767c7af9eec2c04#2ca7b1d6bb24cd05ca30a428efb924268c8dbc51_11689_11690)
>
> Howewer, the commit was nontheless merged with the commit message
>
> > Accept changes to base-exports
> >
> > [...](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9676/diffs?commit_id=3bde9d3da246e29a1d46442bb9b5b962aa917a5b)
>
> In retrospect, that sounds incorrect, doesn't it?

In my defence (since it was me who wrote the GHC MR and the corresponding CLC proposal, and reviewed the API diff) while I did carefully review the API diff (several times, through various tweaks and rebasing), translating from that into "do my changes do what I think they do and is that correct" is not actually that easy. So yes we have tooling that catches unintentional API changes, but interpreting intentional API changes is not so easy.

Here's the API diff:
https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9676/diffs#diff-content-2ca7b1d6bb24cd05ca30a428efb924268c8dbc51

In particular, I think the mistake I made in reviewing the diff was that I was looking to check that I had correctly replaced `IoSubSystem` by `IoManagerFlag`. This included all instances. So in my mind I "correctly" removed instances for `IoSubSystem` and added corresponding instances for `IoManagerFlag` (which is why the new type also has an Enum instance). What I missed is that I should have deviated from that simple replace and _kept_ instances on `IoSubSystem`, while at the same time adding instances for `IoManagerFlag`.
